### PR TITLE
Implemented layout generation in desugaring phase.

### PR DESCRIPTION
**Pull request by mdimeglio**
* Changed INSTALL.md to reflect need for GNU sed for some tests
* Modified Core language `Type t` `Sigils` so that `TCon` sigils are of type `Sigil ()` and `TRecord` sigils are of type `Sigil (DataLayout BitRange)`. This is because in the core language, `TCon` is only used for abstract types, and abstract types have no layout. `TRecords` must always have a layout if they are boxed.
* Added layout generation in the desugaring phase for types whose layout wasn't provided.
* Changed a few test cases to WIP because they involve records containing boxed abstract types, which we will probably be removing from the language, and the initial dargent implementation doesn't support.

### CODE REVIEW

File: cogent/src/Cogent/DataLayout/Desugar.hs [hunk 1]
```diff
@@ -1,27 +1,73 @@
...
+                                    , RepExpr(RepRef, Prim, Offset, Record, Variant)
+                                    )
+import Cogent.DataLayout.TypeCheck  (desugarSize)
 import Cogent.DataLayout.Core
+import Cogent.Core                  ( Type (TUnit, TPrim, TSum, TRecord, TCon, TVar, TVarBang, TFun, TString, TArray, TProduct)
+                                    )
+
+{- * DESUGARING 'Sigil's -}
+
+-- | After normalisation, 'TCon _ _ _' values only represent primitive and abstract types.
```
Comment (zilinc) [inline cogent/src/Cogent/DataLayout/Desugar.hs]: After desugaring, instead of nomalisation.

File: cogent/src/Cogent/DataLayout/Desugar.hs [hunk 2]
```diff
@@ -1,27 +1,73 @@
...
+      -- | This type should be obtained by
+      --   1. Start with the raw type that the `Sigil` is attached to
+      --   2. Replace the top level sigil with `Unboxed`
+      --   3. Desugar the resulting type
+  :: (Type t)
+
+      -- | Since desugarSigil is only called for normalising boxed records (and later, boxed variants),
+      --   the 'Maybe DataLayoutExpr' should always be in the 'Just layout' alternative.
+  -> Sigil (Maybe DataLayoutExpr)
+  -> Sigil (DataLayout BitRange)
```
Comment (zilinc) [inline cogent/src/Cogent/DataLayout/Desugar.hs]: Is it easier and cleaner to implement it as `desugarSigil :: Type t -> Sigil (DataLayout BitRange)` and pattern match on the sigil in each kind of type? IT was a little confusing why you need to break up a type into a type without top-level sigil and the sigil.

**Comment by zilinc** [comment]
Also, if your code break too many tests, what you could do is to add a flag to the compiler to enable the dargent compilation, thus none of the tests in the regression will be affected.

**Comment by mdimeglio** [comment]
This is probably unworkable once I get to code generation because the Dargent code generation will replace the existing code generation for boxed types. Perhaps we could have a way of marking tests as not compatible with Dargent, and a special test suite flag which doesn't run these tests?

File: cogent/src/Cogent/DataLayout/Desugar.hs (hunk 1 above)
Comment (zilinc) [reply cogent/src/Cogent/DataLayout/Desugar.hs]: @mdimeglio  Oh, if you mean the WH-normalisastion in the surface typechecker, instead of A-nomalisation, then you're right.

File: cogent/src/Cogent/DataLayout/Desugar.hs (hunk 2 above)
Comment (zilinc) [reply cogent/src/Cogent/DataLayout/Desugar.hs]: Fair. Maybe it's worth making a comment.

**Comment by zilinc** [comment]
Yes. I'll open a ticket to memorise it.
